### ISSUE TITLE
GPU Support?

### ISSUE THREAD

**Issue by alexbw** [conda-forge/conda-forge.github.io#63]
I've got a couple packages I'm preparing for upload that rely on GPUs. I'm not up to speed on what open-source CI solutions offer, but would building against a VM w/ GPUs be supported? If it would require pitching in or donating to the project, I'm pretty sure I can figure some way to help.

**Comment by pelson** [comment]
I'm completely out of my depth on this one. @msarahan any knowledge on the subject?

**Comment by jakirkham** [comment]
I'm also interested in this. The trick is most CIs do not provide GPUs. However, if you are willing to work with OpenCL, which works with CPUs and GPUs, then we can work together on getting that support on CIs.

**Comment by alexbw** [comment]
Unfortunately, CUDA really is the de-facto standard for machine learning. I
think NVidia is generally interested in helping out the open-source
community, so it might be worth starting a conversation with them about
helping out. I'll test out conda-forge with non-GPU packages, and if things
seem to work smoothly, then can start talking with them.

Question -- is conda and conda-build updated regularly on this system? My
packages are in Lua, and support for them was only added recently.

**Comment by jakirkham** [comment]
I thought you might say that. Unfortunately, without the CI support, we are kind of in a bind on this one. If NVIDIA is willing to work on CI support with GPUs, that would be great.

To be completely honest with you, I don't think we should be holding our breath on this. The real problem is that most CI services are leasing time on other infrastructure. Primarily Google and Amazon's infrastructure. Unless someone has the infrastructure with GPUs that they are willing to lease to some CI service for this purpose, we are kind of stuck. I think we can all imagine what they would prefer to do with this infrastructure, right? However, if you od figure out something on this, please let us know and we can work on something.

I'm guessing you are using Torch then? At a bare minimum, let's work on getting Torch's dependencies in here. At least, that will make your job a little easier, right? For that matter any run-of-the-mill Lua packages that you have would be good to try to get in, as well. It should help you and others looking for more Lua support in conda. How does that sound?

**Comment by jakirkham** [comment]
This repo seems to use NVIDA components for [their CI]( https://github.com/maps-gpu/MAPS/blob/master/travis/install.sh ).

**Comment by jakirkham** [comment]
Did you see the link above, @alexbw?

This might not be totally impossible after all, but I think we should do some research on how this works. What platforms are you trying to support? Just Linux? Mac also? I'm totally out of my depth on Windows. So we may need someone else to show us the ropes there.

**Comment by alexbw** [comment]
Saw the link. Looking more into this, but on the timescale of a few weeks.

**Comment by jakirkham** [comment]
So, I looked a little bit more closely at this and it looks like one could add [GPU libs]( https://developer.nvidia.com/cuda-downloads ) to CentOS 6 (what we are currently using). There is Mac and Windows support too, but IMHO this is secondary to getting Linux up and working. However, I am not seeing any support for CentOS 5 (a platform we were debating switching too), which is something to keep in mind.

**Comment by msarahan** [comment]
Good to know.  We are collecting datapoints on whether continuing with CentOS 5 is a good idea.  If anyone knows of definitive reasons to stay with CentOS5, it is currently preventing:

- Qt5
- complete LLVM
- now, GPU libs

**Comment by jakirkham** [comment]
Glad you say this, @msarahan. Was debating cross referencing, but didn't want to have a mess of links. Are there still many customers using a CentOS 5 equivalent Linux? Could we maybe build the compiler on CentOS 5 and somehow add it to CentOS 6?

**Comment by msarahan** [comment]
Building the compiler on the older architecture doesn't help.  What matters is the GLibc version present on the build system when packages are built.

We don't have hard data on customers, outside of HTTP headers for downloads of packages.  We're digging through that to see how many people have kernels older than the one corresponding to CentOS 6.

**Comment by jakirkham** [comment]
Right. I was just hoping there was a way we could somehow have both. I guess in the worst case some things could be CentOS 6 as needed. Will that have any consequences if we mix the two? Is that already done to some extent (noting that Qt5 was mentioned).

Yeah, it seems like it would be good to give a survey. Might need an incentive to make sure it actually gets filled out.

**Comment by jakirkham** [comment]
Also, interesting footnote (though I would appreciate if other people check and make sure I am reading this right as IANAL), it appears that at least some of the CUDA libraries can be shipped. This means we could create a CUDA package that simply makes sure CUDA is installed in the CI and moves them into the conda build prefix for packaging. The resulting package could then be added as a dependency of anything that requires them (e.g. Torch, Caffe, etc.). This would avoid us having to add these hacks in multiple places and risk losing them when we re-render. Furthermore, we would be able to guarantee that the libraries we used to build would be installed on the user's system.

**Comment by jakirkham** [comment]
We should verify whether one version of CUDA for one Linux distro and version can be used on other Linux distros and versions easily or if we need to have multiple flavors. This survey will have to be extended to other OSes at some point, but starting with Linux makes the most sense to me.

**Comment by jakirkham** [comment]
So, I am trying something out in this PR ( https://github.com/conda-forge/caffe-feedstock/pull/1 ). In it, I am installing CUDA libraries into the docker container and attempting to build against them. It is possible some tests will fail if we don't have access to an NVIDA GPU so we will have to play with that. Also, we don't have cuDNN as that appears to require some registration process that I have not looked into yet and may be a pain to download in batch mode.

In the long run, I expect the CUDA libraries will be wrapped in their own package for installation and packages needing them will simply install these libraries. We may need to features to differentiate different GPU variants (CUDA/OpenCL). However, that CUDA package will probably need to hack the CI script in a similar way.

Definitely am interested in feedback. So, please feel free to share.

**Comment by jakirkham** [comment]
Another thought might be that we don't ship the CUDA libraries. Instead we have a package that merely checks to set that they are installed via a pre or post-link step. If it fails to find them, the install fails. This would avoid figuring out where the libraries can or cannot be distributed safely. Hopefully, as we are linking against the CUDA API. All that will matter is that we have an acceptable version of the CUDA libraries to run regardless of what Linux distribution it was initially built on.

**Comment by jakirkham** [comment]
Appears Circle CI does provide GPU support or at least that is what my [testing suggests]( https://github.com/conda-forge/caffe-feedstock/pull/1#issuecomment-205901214 ).

**Comment by jakirkham** [comment]
Also, as FYI, in case you didn't already know @msarahan, CentOS 5 [maintenance support ends March 2017]( https://wiki.centos.org/FAQ/CentOS5#head-a2fd35a09ebf776661ab0f1288762329bb64209c ). In other words, less than a year. That sounds like a pretty big negative to me. Given how many recipes have been added from conda-recipes and how many remain to be added at this point, trying to add a CentOS 5 switch before that point sounds challenging. Not to mention, we may find ourselves needing to migrate back to CentOS 6 by that point. Maybe it is just me, but I'm starting to feel a lot of friction in switching to CentOS 5. Is it reasonable that we consider just accepting CentOS 6 as part of this transition?

**Comment by kyleabeauchamp** [comment]
FWIW, we have GPU support on Omnia.  Might be worth reading over.

https://github.com/omnia-md/conda-recipes

https://github.com/omnia-md/omnia-build-box

**Comment by jakirkham** [comment]
Thanks for these links @kyleabeauchamp. I'll certainly try to brush up on this.

Do you have any thoughts on this PR ( https://github.com/conda-forge/caffe-feedstock/pull/1 )? Also, how do you guys handle the GPU lib dependency? Is that packaged somehow, used from the system (possibly with some sort of check), some other way?

**Comment by kyleabeauchamp** [comment]
So AFAIK our main use of GPUs was building a simulation engine OpenMM (openmm.org).  OpenMM is a C++ library and can dynamically detect the presence of CUDA support (via shared libraries) at runtime.  This means that we did *not* package and ship anything related to CUDA.  We basically just needed CUDA on the build box to build the conda package, then let OpenMM handle things later dynamically.

**Comment by kyleabeauchamp** [comment]
Looks like our dockerfile is somewhat similar to your CUDA handling:

https://github.com/omnia-md/omnia-build-box/blob/master/Dockerfile#L25

**Comment by jakirkham** [comment]
Ah, ok, thanks for clarifying.

The trick with Caffe, in particular, is that it can use CPU, CUDA, or OpenCL. CPU support is always present; however, a BLAS is required, which includes a user's CPU choice (OpenBLAS, ATLAS, MKL, or possibly some hack to add other options) and a GPU choice (if any) cuBLAS or ViennaCL. Thus, having this dynamically determined ends up not really being as nice as it could be. To allow actual selection will require feature support and possibly multiple rebuilds of Caffe.

One simple route might be to just always use ViennaCL, which can abstract the difference between the OpenCL and CUDA options. Also, it can always fallback to the CPU if no GPU support is present. Though I expect this layer of abstraction comes at some penalty, the question is how severe is that penalty. Would a solution like this work with OpenMM? I don't know if its GPU support proceeds primarily through a GPU BLAS or some other mechanisms. For instance, is it using FFTs?

If you have deep learning interests, this may be relevant. In the case of Caffe, it can optionally support cuDNN. Researchers will want this support out of the box. Not only is this tricky because it may not be provided for due to hardware or software reasons, it is tricky because downloading cuDNN requires a registration step with unclear licensing restrictions. One way we might resolve this is to request cuDNN be loaded on an appropriate Docker container. NVIDA does do this with Ubuntu 14. However, I don't see a similar container for CentOS 6 and am unclear on whether it would be a supported platform. Ultimately, it will require us to communicate with NVIDA at some point to see what we need to do here to stay above board while providing users state of the art support.

Fortunately, NVIDA is very clear as to what parts of the CUDA libraries can be distributed down to file level of what can and cannot be distributed. So, the concerns with cuDNN do not affect this.

**Comment by jakirkham** [comment]
Another thought for more versatile support would be to use the [clMath]( https://github.com/clMathLibraries ) libraries.

**Comment by kyleabeauchamp** [comment]
OpenMM dynamically chooses the best platform at runtime, with options including CPU (SSE), CUDA, OpenCL, and CPU (no SSE / reference).  It does use FFTs.  The idea with OpenMM is to build and ship binaries that support all possible platforms, then select at runtime.

**Comment by Lnaden** [comment]
Continuing the discussion from https://github.com/conda-forge/staged-recipes/issues/299#issuecomment-297404016, the problem we have is that OpenMM requires the CUDA toolkit libraries and AMD SDK to build its GPU features against. Moving OpenMM over to conda-forge is a large barrier to moving the remainder of omnia's packages over.

As @kyleabeauchamp mentioned on here, OpenMM does not actually ship CUDA or the AMD SDK, it only uses the libraries to build its own GPU components to ship out and accesses user's local GPU software. OpenMM requires CUDA Toolkit 8.0 and AMD SDK 3.0 for its current builds.

As discussed by @jakirkham and @jjhelmus on the linked issue, there are license problems if we tried to ship the toolkits with OpenMM, but I'm not sure that would be required. 

Tagging @peastman, @jchodera 
xref: https://github.com/conda-forge/staged-recipes/issues/299#issuecomment-297404016
xref: https://github.com/conda-forge/staged-recipes/issues/299#issuecomment-297407579
xref: https://github.com/conda-forge/staged-recipes/issues/299#issuecomment-297409444

**Comment by scopatz** [comment]
> Fortunately, NVIDA is very clear as to what parts of the CUDA libraries can be distributed down to file level of what can and cannot be distributed

@jakirkham - what do you mean by this? Is there a file listing somewhere?

**Comment by jjhelmus** [comment]
NVIDIA lists the files which are redistributable under their license are given in Attachment A of the EULA.  The document moves around with new releases of CUDA but currently is available [here](https://docs.nvidia.com/cuda/eula/index.html#attachment-a).

**Comment by isuruf** [comment]
You still have the line

>  You agree to defend, indemnify and hold harmless NVIDIA and its Affiliates, and their respective employees, contractors, agents, officers and directors, from and against any and all claims, damages, obligations, losses, liabilities, costs or debt, fines, restitutions and expenses (including but not limited to attorney’s fees and costs incident to establishing the right of indemnification) arising out of or related to you and your Enterprise, and their respective employees, contractors, agents, distributors, resellers, end users, officers and directors use of Licensed Software outside of the scope of the AGREEMENT or any other breach of the terms of the AGREEMENT. 

in that link which was the exact same reason we don't ship MKL and Intel runtime libraries.

**Comment by isuruf** [comment]
I think it's saying we will defend and indemnify NVIDIA if a conda-forge user sues them. IANAL though

**Comment by scopatz** [comment]
Also, the sentence before that is specifically about share-alike linking, "you agree that you will not (nor authorize third parties to, [...] use the Licensed Software in any manner that would cause the Licensed Software to become subject to an Open Source License. Nothing in the AGREEMENT shall be construed to give you a right to use, or otherwise obtain access to, any source code from which the Software or any portion thereof is compiled or interpreted."

So this seems to be about if someone tries to GPL their code. 

Since the redistributable white-list only includes binaries and headers (not source code), I am not sure how we could be at fault because there is no oppurtunity for a cudatoolkit package to include source in the first place.

**Comment by scopatz** [comment]
I thought MKL's license didn't even allow you to redistribute binaries.

**Comment by scopatz** [comment]
Thanks for the whitelist @jjhelmus!

**Comment by dougalsutherland** [comment]
> Since the redistributable white-list only includes binaries and headers (not source code), I am not sure how we could be at fault because there is no oppurtunity for a cudatoolkit package to include source in the first place.

I think it's [somewhat controversial](https://en.wikipedia.org/wiki/GNU_General_Public_License#Linking_and_derived_works) whether distributing GPL software linked to non-free binaries is okay; that's probably what they mean.

**Comment by jakirkham** [comment]
> I thought MKL's license didn't even allow you to redistribute binaries.

One is allowed to redistribute MKL as long as their license is honored.

ref: https://software.intel.com/en-us/license/intel-simplified-software-license

**Comment by jchodera** [comment]
Is the concern here (1) whether hosting `linux-anvil` derivatives that contain install CUDA toolkits to allow the automated building of packages that link against CUDA or OpenCL is permissible, (2) whether distribution of conda packages compiled against CUDA libraries is permissible, or (3) whether the CUDA toolkit could itself be redistributed as a conda-forge package?

**Comment by scopatz** [comment]
For me, the concern is (3)

**Comment by jakirkham** [comment]
Personally would say 2+3 are concerns. 3 less so as we could just use the package from `defaults`.

**Comment by dougalsutherland** [comment]
Yeah, Anaconda has special permission (IIRC) from Nvidia to redistribute `cudatoolkit` on `defaults` and we can just rely on that. I think (1) and (2) are both potential concerns, along with the practical issue of not having GPUs on the CI services to test packages.

**Comment by jchodera** [comment]
> Yeah, Anaconda has special permission (IIRC) from Nvidia to redistribute cudatoolkit on defaults and we can just rely on that.

It looks like [those packages](https://anaconda.org/anaconda/cudatoolkit/files) are not kept up to date. Both CUDA 9.1 and 9.2 have been released since (along with some patches), but those do not seem to be available there.

**Comment by jakirkham** [comment]
Please raise with Anaconda. Thanks:)

**Comment by jchodera** [comment]
cc: https://github.com/ContinuumIO/anaconda-recipes/issues/140

**Comment by scopatz** [comment]
> Yeah, Anaconda has special permission (IIRC) from Nvidia to redistribute cudatoolkit on defaults

So I don't think that is true, per se. As @jakirkham, @jjhelmus, and others have brought up, parts of cudatoolkit are freely redistributable according to the EULA.  These freely redistributable portions are all that the `defaults` package contains. 

> and we can just rely on that

As @jchodera points out, these packages are not kept up-to-date.  To that end, there is now PR to add a similar recipe to what's on `defaults` (freely redistributable)  into staged recipes at conda-forge/staged-recipes#6240.

> I think it's somewhat controversial whether distributing GPL software linked to non-free binaries is okay; 

This seems like it would only be an issue for any other packages (ie not cudatoolkit) that happen to be GPL and happen to link to CUDA. This should not affect the distrubtion of cudatoolkit itself.

**Comment by isuruf** [comment]
> Also, the sentence before that is specifically about share-alike linking

> So this seems to be about if someone tries to GPL their code.

I think the sentence I mentioned is separate from the whole `you agree that you will not (nor authorize third parties to):` part.

**Comment by scopatz** [comment]
@isuruf - so the sentence you mentioned (`You agree to defend, indemnify ...`) is in the same clause (1.2.1) as the `you agree that you will not: ...`. So I think it is fair to read them as related propositions. 

Additionally, we offer conda-forge under BSD-3-Clause. This means that all of conda-forge's repos, packages, etc are provided by us without any "express or implied warranties."  So while we may be obligated to `defend, indemnify, ...` NVIDIA from our users, our users have no grounds to make a claim against us or NVIDIA.  Our BSD license serves as protection for NVIDIA.

**Comment by xmnlab** [comment]
should we try to invite some person from nvidia to help us in this issue?

**Comment by jakirkham** [comment]
> I sent oss-requests@nvidia.com an email asking for clairifacations. Worst case I could request the source code and try to compile it.

@scopatz, did you ever hear back?

**Comment by scopatz** [comment]
@jakirkham - sadly not yet!

**Comment by jakirkham** [comment]
Hmm...is there someone or someones from NVIDIA at SciPy? Maybe we can find a better contact than the public facing generic OSS support email.

**Comment by jchodera** [comment]
[Mark Berger](https://www.linkedin.com/in/1markberger/) is the Senior Alliance Manager for Life & Material Sciences at NVIDIA, and has always been tremendously helpful to us in connecting up with the right folks inside the organization to troubleshoot. I'd connect up with him if you have any trouble finding the right person inside of NVIDIA.

**Comment by hadim** [comment]
Any news on this? Having cuda as a conda package would be great!

**Comment by scopatz** [comment]
Yes, unfortunately, our legal counsel (via NumFOCUS) informed us that according to the cudatoolkit EULA, we are *not* allowed to redistribute cudatoolkit.  We'll need to get special permission to redistribute these packages. I'll be attempting to pursue this with them.

**Comment by hadim** [comment]
That's suck... But thank you for keep trying @scopatz and others. Let us know here.

**Comment by scopatz** [comment]
No problem!  I am sure having other people ask NVIDIA for this too, can only help our case :)

**Comment by isuruf** [comment]
> We'll need to get special permission to redistribute these packages.

Given that we are okay with the indemnify clause, do we need special permission if we use cudatoolkit from defaults and use that to build packages?

**Comment by scopatz** [comment]
Yes, I think so. Anaconda Inc is who assumes all legal risk from the defaults packages

**Comment by alexbw** [comment]
Hey all, it looks like PyTorch (very popular ML package) copies CUDA .so files directly into site-packages. Those folks might know about licensing. https://github.com/pytorch/pytorch/blob/2681852438f9f80b43fd2502d83839140733b68c/conda/integrated/build.sh#L61-L119

**Comment by scopatz** [comment]
@alexbw - they either are breaking the license or (more likely) have another license from NVIDIA to do so.

**Comment by dgasmith** [comment]
@scopatz We could potentially poke Nvidia about this as well from a fairly large userbase that is often featured in their advertisements (molecular dynamics). Do you already have a contact within the company or should we reach out to our local Nvidia representative?

@jchodera @Lnaden

**Comment by jchodera** [comment]
@dgasmith: We have found [Mark Berger](https://www.linkedin.com/in/1markberger), the Senior Alliance Manager, Life & Material Sciences, to be incredibly helpful. He could likely put you in touch directly with the people necessary to grant license exceptions. I'd be happy to introduce you via email.

**Comment by dgasmith** [comment]
Mark is exactly who was on my mind :)

I do want to know who at Nvidia has already been contacted so that the right folks can get synced together and not have too many loose efforts.

**Comment by hmaarrfk** [comment]
@scopatz did the legal council state what clause made the cuda toolkit non-redistributable?

Are we allowed to link to cuda stuff so long as we pull in the dependency from defaults?

Ref: https://docs.nvidia.com/cuda/eula/index.html#attachment-a

**Comment by scopatz** [comment]
Basically, counsel has said that the EULA - even attachment A - only refers to applications (it does), and conda-forge cannot be considered an application under any reasonable definition.


Also, it looks like they have added the following to attachment A, which we don't meet:

```
The NVIDIA CUDA Driver Libraries are only distributable in applications that meet this criteria:

1. The application was developed starting from a NVIDIA CUDA container obtained from Docker Hub or the NVIDIA GPU Cloud, and
2. The resulting application is packaged as a Docker container and distributed to users on Docker Hub or the NVIDIA GPU Cloud only.
```

**Comment by hmaarrfk** [comment]
I guess we can't distribute the Cuda driver. Many people are OK with installing that themselves :/.

Second question would be:
Is linking to the libraries on defaults OK by conda-forge standards?

**Comment by scopatz** [comment]
Linking to the deafults libraries is fine for us

**Comment by hadim** [comment]
Then we can say goodbye to reproducible workflow involving CUDA. This is really a non sense we can't ship it in conda-forge.

I understand the legal aspect of it. What I don't understand is why they took that decision...(unless it's a misunderstanding and we can actually ship it conda?)

**Comment by scopatz** [comment]
> Then we can say goodbye to reproducible workflow involving CUDA. This is really a non sense we can't ship it in conda-forge.

> I understand the legal aspect of it. What I don't understand is why they took that decision.

I agree on all counts.  

> (unless it's a misunderstanding and we can actually ship it conda?)

From speaking with NVIDIA, I am reasonably sure that this is not a misunderstanding on their part. The folks who work on ML and cuda and PyData stuff at NVIDIA are different than those in their legal department. I don't know the justification for the license as it stands, but we have to live with it until:

1. they change the EULA
2. they give us a written exception, which I have asked for but they haven't responded yes or no

**Comment by hadim** [comment]
Thank you @scopatz for working on all those things.

In the meantime for those who want to have reproducible CUDA installation (only Linux unfortunately), see the following gist: https://gist.github.com/hadim/a4fe638587ef0d7baea01e005523e425

**Comment by ericpre** [comment]
Now that building package with CUDA is working fine on linux (in case of https://github.com/conda-forge/prismatic_split-feedstock, it went smoothly), how would it be possible to add windows support? Is it worth trying to add a `cudatoolkit-dev` package for windows in order to build GPU packages on windows?

**Comment by isuruf** [comment]
PRs welcome. We could use `cudatoolkit-dev` in Linux too to avoid using /usr/include because NVIDIA images have been polluting the includes into that folder.

**Comment by ericpre** [comment]
Ok, I will give a try.

**Comment by jakirkham** [comment]
@ericpre, for an example of how to build GPU packages would take a look at [nvidia-apex]( https://github.com/conda-forge/nvidia-apex-feedstock ) as a simple example. Sorry there are no docs atm.

**Comment by isuruf** [comment]
@jakirkham, does `nvidia-apex-feedstock` build windows cuda packages?

**Comment by znmeb** [comment]
I've got an NVIDIA Jetson AGX Xavier (8-core Tegra `aarch64` CPU plus a 512-core Volta GPU) that I can test on. I just discovered Miniforge a couple of days ago - it appears to be working out of the box on the Tegra CPU.

I also have a laptop with a GTX 1050 Ti GPU / x86_64 CPU but there's plenty of other ways to get at the GPU there.

**Comment by tyler274** [comment]
Why isn't OpenCL included in OpenCV already. CI support and license issues aren't an issue so why is it blocked on Cuda?

**Comment by mlkmooeeeed** [comment]
[![Repo Freeze Check](https://github.com/github/docs/actions/workflows/repo-freeze-check.yml/badge.svg?branch=dependabot%2Fdocker%2Fnode-19.1.0-alpine)](https://github.com/github/docs/actions/workflows/repo-freeze-check.yml)
